userspace/libsinsp/container_engine/cri.cpp

```diff
@@ -243,7 +243,32 @@ bool cri::resolve(sinsp_threadinfo *tinfo, bool query_os_for_missing_info)
 			libsinsp_logger()->format(sinsp_logger::SEV_DEBUG,
 					"cri_async (%s): Starting synchronous lookup",
 					container_id.c_str());
+			// `lookup_sync` function directly invokes the container engine specific parser `parse`
 			done = m_async_source->lookup_sync(key, result);
+			if(!result.m_image.empty() || result.is_pod_sandbox())
+			{
+				/*
+				* Only for synchronous lookup option (e.g. Falco's default is async not sync)
+				*
+				* Explicitly check for the most crucial retrieved value (`m_image`) to be present before enabling the 
+				* fast-track container add option. At this point, the container with only the cgroup (container id) was 
+				* already added to the cache. Therefore, we can proceed to call `replace_container`.
+				* 
+				* Bypassing the round-trip process:
+				* `source_callback` -> `notify_new_container` ->
+				* `container_to_sinsp_event(container_to_json(container_info), ...)` ->
+				* `parse_container_json_evt` -> `m_inspector->m_container_manager.add_container()`
+				*
+				* In `parse_container_json_evt`, we still re-add the container to support native 'container' events
+				* and new container callbacks that may expect the container as JSON in the artificial sinsp evt.
+				* However, we can avoid delays by storing the container struct in the container cache now. 
+				* This is beneficial because syscall events do not explicitly require container events, instead, 
+				* they directly retrieve container details from the container cache. This new feature can mitigate 
+				* issues noted by adopters, such as the absence of container images in syscall events even when 
+				* disabling async lookups.
+				*/
+				cache->replace_container(std::make_shared<sinsp_container_info>(result));
```

Comment:
Thanks @therealbobo follow up question:

Where is the status changed the function? It only performs the following:
```
void sinsp_container_manager::replace_container(const sinsp_container_info::ptr_t& container_info)
{
	auto containers = m_containers.lock();
	ASSERT(containers->find(container_info->m_id) != containers->end());
	(*containers)[container_info->m_id] = container_info;
}
```

Solely the `add_container` sets the current status, but also doesn't change it to successful either it just sets the current status.

What is the meaning of CACHED state? Is it set to this status when we first add the container to the cache when we parsed the crgoup? In that case it already has that CACHED status. I can check more, just wanted to ask in case you are still online right now.

Reply:
The problem is that, until now, the container_info was not stored in the cache. Doing so we are changing the behaviour of `parse_container_json_evt`. In particular
```
void sinsp_parser::parse_container_json_evt(sinsp_evt *evt)
{
	if(evt->get_tinfo_ref() != nullptr)
	{
		const auto& container_id = evt->get_tinfo_ref()->m_container_id;
		const auto container = m_inspector->m_container_manager.get_container(container_id);
		if(container != nullptr && container->is_successful())
		{
			SINSP_DEBUG("Ignoring container event for already successful lookup of %s", container_id.c_str());
			evt->set_filtered_out(true);
			return;
		}
	}
        ...
```
With your patch the `const auto container = m_inspector->m_container_manager.get_container(container_id);` lookup will result in getting a container_info with a `SUCCESSFUL` state and, therefore, the event will not be parsed due to the `if(container != nullptr && container->is_successful())` check. So, as you pointed out
> Is it set to this status when we first add the container to the cache when we parsed the cgroup?
the `CACHED` state is just an artificial state to know when is the first time that we see _that_ container. I hope that I was more clear. Feel free to reach out if I wasn't! 😄

Reply:
@therealbobo first of all thanks for caching this! Looking a bit more carefully now, I think ... I don't even know how to say it. There is an overall improvement opportunity :sweat_smile: ...

 Perhaps the root cause actually is that we init the sinsp_container_lookup state as successful in the `sinsp_container_lookup` class.

`state m_state = state::SUCCESSFUL;` -- why would we do this (?!), it seems just wrong (!!!!!) If anything plz init with FAILED ...

Current enum

```
enum class state
	{
		STARTED = 0,
		SUCCESSFUL = 1,
		FAILED = 2
	};
```

I initially overread you suggesting adding an entire new state, and thought `CACHED` already existed, but I never recalled seeing it.

We should maybe add 2 new states
- `EMPTY` or similar -- a state that indicates nothing has happened yet at init time of the class. I don't like at all that we trust to always override with `sinsp_container_lookup::state::STARTED`, evidentially it is easy to break, that's why no one dares to touch and optimize these code paths ... or we stick to the existing state FAILED for the init.
-  Instead of CACHED, maybe call it `API_LOOKUP_SUCCESSFUL` -  It may be more closely related to the SUCCESSFUL term? Or we stick to the existing state STARTED.

Not sure if we rely on the actual enum values, but I can just add the new states at the end, I believe it doesn't matter - or if it matters then it can only be because of scap files as always.

Reply:
Done. I'll open a separate PR to fix the init state given how easily things can be broken and merging this is more important. But we still need to fix the container engine up all around, else the next person makes the same mistakes.

Reply:
Also rebased.

Reply:
TBH I like `CACHED` a lot (though I'm not going to bikeshed over the name _too_ much ;]). `STARTED` is completely wrong here as it's meant to indicate a lookup is happening in the background (we know it isn't, it has just finished successfully).

Introducing another state (even if it happens to be handled the same as an already existing one) doesn't cost anything and adds more clarity IMO.

Reply:
Ok yeah that is easy to add, no problem doing it.

Reply:
Given we have https://github.com/falcosecurity/libs/issues/1708 we could also defer adding new states until a proper refactor.